src/dockerfile.rs

```diff
@@ -0,0 +1,373 @@
+use lazy_static::lazy_static;
+use regex::{Regex, RegexSetBuilder};
+use std::{collections::HashSet, error::Error, fs, path};
+
+lazy_static! {
+    static ref FROM_DIRECTIVE: regex::Regex =
+        Regex::new(r"(?i)^FROM\s(?P<image>[^\s]+)?(?:\sAS\s)?(?P<stage>[\w+]+)?").unwrap();
+    static ref VERSION_REGEX: &'static str = r"v?([0-9]+(?:(?:\.[a-z0-9]+)|(?:-(?:kb)?[0-9]+))*)";
+    static ref TAG_REGEX: regex::RegexSet = RegexSetBuilder::new(&[
+        format!(r"^{}(?P<suffix>-[a-z0-9.\-]+)?$", *VERSION_REGEX),
+        format!(r"^(?P<prefix>[a-z0-9.\-]+-)?{}$", *VERSION_REGEX),
+        format!(
+            r"^(?P<prefix>[a-z\-]+-)?{}(?P<suffix>-[a-z\-]+)?$",
+            *VERSION_REGEX
+        ),
+    ])
+    .case_insensitive(true)
+    .build()
+    .unwrap();
+}
+
+pub struct Dockerfile {
+    pub path: String,
+    pub content: String,
+}
+
+impl Dockerfile {
+    pub fn new(path: path::PathBuf) -> Result<Self, Box<dyn Error>> {
+        let dockerfile = Self {
+            path: String::from(path.to_str().unwrap()),
+            content: fs::read_to_string(path)?,
+        };
+        Ok(dockerfile)
+    }
+
+    fn images(&self) -> Vec<Image> {
```

Comment:
When should I return a reference (`Vec<&Image>`) instead of a value?

Reply:
As with everything, it depends what happens next, like if youre happy to pass on the actual value within the `move` then its fine, if you need to keep the reference to it elsewhere, then you would pass around the reference instead.

Reply:
Yeah, it really depends on your needs. In this case for example, you probably don't want to return the "reference", ie you build and own it within the function itself, so once the function returns, it will be deallocated.

It's quite nuanced and it depends on what you want to do - my view on this is: don't think of references here in the same way you think of pointers in C *and* you're probably better off (even performance wise) doing the simplest thing and letting the rust compiler optimise. Failing that, then you can look at these things.

Reply:
As a rule of thumb, I'd return references wherever I can.

In this instance, you can't return a reference, as the `Image` gets constructed as part of this function, so we need to transfer ownership of it to the caller.

Reply:
yeah i think i'm still thinking of references in the same way as pointers in C and Go 😅but the advice here is helpful. @matt-thomson, does that mean when constructing a struct, i should generally return `&Self` instead of `Self`?